#! /bin/sh /usr/share/dpatch/dpatch-run
## fix_ppc_build_using_gcc-4.4_r131472.dpatch by Mirco Bauer <meebey@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Patch taken from upstream's SVN, revision: r131472
## DP: 2009-04-10  Zoltan Varga  <vargaz@gmail.com>
## DP: * include/private/gc_locks.h (GC_test_and_set): Merge a change from libgc
## DP:   libatomic-ops to fix the ppc build with gcc 4.4.

@DPATCH@
diff -urNad mono-2.4+dfsg~/libgc/include/private/gc_locks.h mono-2.4+dfsg/libgc/include/private/gc_locks.h
--- mono-2.4+dfsg~/libgc/include/private/gc_locks.h	2009-02-14 00:32:31.000000000 +0100
+++ mono-2.4+dfsg/libgc/include/private/gc_locks.h	2009-06-13 14:14:44.000000000 +0200
@@ -160,15 +160,15 @@
           int temp = 1; /* locked value */
 
           __asm__ __volatile__(
-               "1:\tlwarx %0,0,%3\n"   /* load and reserve               */
+               "1:\tlwarx %0,0,%1\n"   /* load and reserve               */
                "\tcmpwi %0, 0\n"       /* if load is                     */
                "\tbne 2f\n"            /*   non-zero, return already set */
                "\tstwcx. %2,0,%1\n"    /* else store conditional         */
                "\tbne- 1b\n"           /* retry if lost reservation      */
                "\tsync\n"              /* import barrier                 */
                "2:\t\n"                /* oldval is zero if we set       */
-              : "=&r"(oldval), "=p"(addr)
-              : "r"(temp), "1"(addr)
+              : "=&r"(oldval)
+              : "r"(addr), "r"(temp)
               : "cr0","memory");
           return oldval;
         }
y